File: extension-gems/jedna-tournaments/examples/simple_agent.py
#!/usr/bin/env python3
import json
import sys
from collections import Counter

class SimpleAgent:
    def run(self):
        while True:
            try:
                line = input()
                data = json.loads(line)
                
                if data['type'] == 'request_action':
                    action = self.decide_action(data['state'])
                    print(json.dumps(action))
                    sys.stdout.flush()
                elif data['type'] == 'game_end':
                    break
                    
            except EOFError:
                break
    
    def decide_action(self, state):
        # Simple strategy: play first playable card, otherwise draw
        if state.get('playable_cards') and len(state['playable_cards']) > 0:
            card = state['playable_cards'][0]
            action = {'action': 'play', 'card': card}
            
            # Add color for wild cards
            if card == 'w' or card.startswith('wd'):
                # Pick the color we have most of
                colors = [c[0] for c in state['hand'] if c[0] != 'w']
                if colors:
                    color_counts = Counter(colors)
                    best_color = color_counts.most_common(1)[0][0]
                else:
                    best_color = 'r'
                
                color_map = {'r': 'red', 'b': 'blue', 'g': 'green', 'y': 'yellow'}
                action['wild_color'] = color_map[best_color]
            
            return action
        elif 'draw' in state.get('available_actions', []):
            return {'action': 'draw'}
        else:
            return {'action': 'pass'}

if __name__ == '__main__':
    SimpleAgent().run()

File: extension-gems/jedna-tournaments/examples/simple_agent.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'

# Basic Jedna agent that plays the first available card
class SimpleAgent
  COLOR_NAMES = {
    'r' => 'red',
    'b' => 'blue',
    'g' => 'green',
    'y' => 'yellow'
  }.freeze

  def run
    loop do
      input = gets
      break if input.nil?

      process_message(JSON.parse(input))
    end
  end

  private

  def process_message(data)
    case data['type']
    when 'request_action'
      respond_with_action(decide_action(data['state']))
    when 'game_end'
      exit
    end
  end

  def respond_with_action(action)
    puts JSON.generate(action)
    $stdout.flush
  end

  def decide_action(state)
    if state['playable_cards']&.any?
      play_card(state['playable_cards'].first, state['hand'])
    elsif state['available_actions']&.include?('draw')
      { 'action' => 'draw' }
    else
      { 'action' => 'pass' }
    end
  end

  def play_card(card, hand)
    action = { 'action' => 'play', 'card' => card }
    action.merge!(wild_color_choice(card, hand)) if wild_card?(card)
    action
  end

  def wild_card?(card)
    card == 'w' || card.start_with?('wd')
  end

  def wild_color_choice(_card, hand)
    best_color = most_common_color(hand)
    { 'wild_color' => COLOR_NAMES.fetch(best_color, 'red') }
  end

  def most_common_color(hand)
    colors = hand.map { |c| c[0] }.reject { |c| c == 'w' }
    color_counts = colors.tally
    color_counts.max_by { |_, count| count }&.first || 'r'
  end
end

SimpleAgent.new.run if __FILE__ == $PROGRAM_NAME
